### PR TITLE
Flamegraph: Move to package

### PR DESCRIPTION
**Pull request by aocenas**
Moving the Flamegraph code into a separate package so it can bu published in NPM and used outside of Grafana. Main usecase right now is the [Pyroscope web app](https://github.com/grafana/pyroscope/tree/main/public/app):

Screenshot 2023-09-05 at 22 07 37

See the draft [PR](https://github.com/grafana/pyroscope/pull/2346) in the pyro web app for usage.

Main part of the changes here are obviously that the code of the flamegraph was moved from the core grafana to a separate `package/` directory.

There are also some other notable changes I will also try to highlight in the code:
- Removed dependency on `grafana/runtime`. This does not make sense in a component package anyway and was used mainly for interaction reporting. This was moved to an injectable hooks that fire on specific interactions.
- Add `getTheme` prop so theme can be injected instead of relying on `useTheme` and it's access to theme in react context.
- `stickyHeader` prop to control the header css. Before it relied on knowing which part of Grafana was this used in which again does not make sense in a package.
- `extraHeaderElements` + `vertical` props. Allows more customization for usecases inside the pyroscope app.

### Wider context

**why new package?**
We want to unify the flamegraph components across grafana and Pyroscope web app. Because of that we need to be able to import the component into non grafana envirnment.

**why not adding it to grafana/ui**
The grafana/ui is probably wider than it should be. Adding visualizations would make it too big to act as a general component library. Visualizations also mostly act as a small self contained applications so may not fit the same abstraction. Also we want to eventually have the flexibility of having separate release for visualizations panel plugins and if they depended on grafana/ui with most of their code they would still be tied to it.

### Testing

The basic test would be to see if the flamegraph still works ok inside grafana either with devenv pyroscope block or using test data source.

To see it used in Pyroscope app this is a bit harder right now. You have to setup local npm server and publish the package there. See the [packages readme](https://github.com/grafana/grafana/blob/77bf8215d4c00791f61f263bfb565d8cdbc63cb1/packages/README.md#L71) for instructions, then you can use the Pyroscope [PR](https://github.com/grafana/pyroscope/pull/2346), install it there and try it out.

### CODE REVIEW

**Comment by tolzhabayev** [comment]
Shouldn't this live in `grafana/ui` or `grafana/experimental`? What is the need for another package?

**Comment by torkelo** [comment]
can't it be used via PanelRenderer or scenes VizPanel?

**Comment by aocenas** [comment]
Sorry for late reply:
@tolzhabayev
> Shouldn't this live in grafana/ui or grafana/experimental? What is the need for another package?

I talked to front end platform and they were quite strongly against adding more visualization into grafana/ui package. Also if we want to speed up the development of plugins we probably want to make them less dependant on bigger packages like grafana/ui which may have less frequent releases.

@torkelo 
> can't it be used via PanelRenderer or scenes VizPanel?

This should be used both in app plugin but also in separate pyroscope web app (and maybe in flamegraph.com) which do not have any grafana runtime stuff.

File: packages/grafana-ui/src/components/Icon/cached.json
Comment (academo) [inline packages/grafana-ui/src/components/Icon/cached.json:4]: unless this icons are directly required on a first page load it is better not to add them to this list as it'll increase the js bundle without a good reason.

**Comment by academo** [comment]
@aocenas if I get this right, you are exporting a single panel (rather complex though) as a single package so it can be used in other apps?

There are some questions I have about this approach:

* Could we do with exporting flamegraph via `@grafana/ui` which is already the place to import UI elements?
* Are we perhaps over-exposing in this new package? Are you 100% sure that every single exported symbol in this package is something you want to have usable in an third party app? Do we have the correct types, documentation and guides? Consider this will be not only used by your team but anyone that would like to write a plugin.
* Is the documentation up to date? Should I need to use this package, could I  use it with the information currently in the readme?
* How is this package published?

**Comment by aocenas** [comment]
@academo 

> Could we do with exporting flamegraph via @grafana/ui which is already the place to import UI elements?

I was told we don't want more visualizations inside the grafana/ui (and I agree). I think @joshhunt can elaborate on this.

> Are we perhaps over-exposing in this new package?

Anything specific you have in mind? In exports single component so not sure how it can expose less.

> Is the documentation up to date?

Again anything specific? I am not saying it has the best readme there is but considering it's a single component I think it has the basics. There is additional work for data plane contracts done in parallel which probably should be linked so that the data requirement is linked.

> How is this package published?

Not sure. Hoped it would be the same as other packages right now mainly to save time. I have to check whether we can do something about the versioning as I am not sure if it makes sense to version the same as grafana.

**Comment by aocenas** [comment]
@academo sorry just now pushed the readme

**Comment by tolzhabayev** [comment]
@aocenas I would love to have a clear why not `grafana/ui` before we merge anything (looks like you pinged Josh above).

And @staton-hyse11 would be interesting to hear what you think?

Additionally please add your squad as owners for this package after this line https://github.com/grafana/grafana/blob/433c76f7f66c8e16130d187f6f6920bc5abd95be/.github/CODEOWNERS#L325

Hooking into publishing pipeline should be done by frontend platform I guess, probably some CI scripts need to be adjusted.

File: packages/grafana-ui/src/components/Icon/cached.json
Comment (academo) [reply packages/grafana-ui/src/components/Icon/cached.json:4]: >They don't load at all if they are not there. I mean if you use grafana/ui outside of grafana you just have to bundle them.

True. I thought this was going to be used only inside the grafana context.

**Comment by aocenas** [comment]
@tolzhabayev I added some more context to the description in this PR

**Comment by aocenas** [comment]
The question of putting this into grafana/ui was discussed separately and we agreed to continue with the PR as it is.

File: public/app/features/explore/FlameGraph/FlameGraphExploreContainer.tsx
Comment (aocenas) [inline public/app/features/explore/FlameGraph/FlameGraphExploreContainer.tsx:30]: To remove grafana/runtime the interaction reporting is injected like this now.

File: public/app/features/explore/FlameGraph/FlameGraphExploreContainer.tsx
Comment (aocenas) [inline public/app/features/explore/FlameGraph/FlameGraphExploreContainer.tsx:29]: Injected theme so we don't rely on having it in the react context

File: packages/grafana-flamegraph/src/FlameGraphHeader.tsx
Comment (aocenas) [inline packages/grafana-flamegraph/src/FlameGraphHeader.tsx:118]: Extra elements rendered here. In pyroscope used to inject a button to download the profile in various formats.

**Comment by joey-grafana** [comment]
Works locally with testdb & pyroscope/phlare backend, but do you have more details to test this in the app?

File: packages/grafana-flamegraph/src/FlameGraphContainer.tsx
Comment (mckn) [inline packages/grafana-flamegraph/src/FlameGraphContainer.tsx:14]: What do you think about also exporting this type? If someone for some reason would like to wrap the Graph in a HOC or similar it could be useful to be able to import the Props type.

**Review by mckn — APPROVED** [review]
LGTM! I really like the work you have done here and I would love to see more of our "bigger" visualisations be structured like this.

File: packages/grafana-flamegraph/src/FlameGraphContainer.tsx
Comment (aocenas) [reply packages/grafana-flamegraph/src/FlameGraphContainer.tsx:14]: Ah yeah will do, definitely makes sense to export

File: public/app/features/explore/FlameGraph/FlameGraphExploreContainer.tsx
Comment (aocenas) [reply public/app/features/explore/FlameGraph/FlameGraphExploreContainer.tsx:29]: Yes in the pyroscope web app.

**Review by adrapereira — APPROVED** [review]

**Pull request merged**
